src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs

```diff
@@ -203,6 +212,88 @@ static bool collectTypeParameters(TypeSymbol type, PooledHashSet<TypeParameterSy
 
                 return declaredUnderlyingType;
             }
+
+            static bool hasSelfReference(TypeSymbol underlyingType, SourceExtensionTypeSymbol definition, ConsList<TypeSymbol>? basesBeingResolved)
+            {
+                Debug.Assert(definition.IsDefinition);
+                PooledHashSet<TypeSymbol> alreadyVisited = PooledHashSet<TypeSymbol>.GetInstance();
+                var result = foundSelfReferenceInErasure(underlyingType, definition, alreadyVisited: alreadyVisited, basesBeingResolved: basesBeingResolved, isContainer: false);
+                alreadyVisited.Free();
+                return result;
+            }
+
+            // Returns true if any erasure in the visited type is the given definition
+            static bool foundSelfReferenceInErasure(TypeSymbol type, SourceExtensionTypeSymbol definition, PooledHashSet<TypeSymbol> alreadyVisited, ConsList<TypeSymbol>? basesBeingResolved, bool isContainer = false)
```

Comment:
>foundSelfReferenceInErasure [](http://example.com/codeflow?start=24&length=27)

This looks unnecessary complicated. What advantage do we get from bundling two unrelated checks? It feels to me that the check could be performed later and with much simpler ```VisitType``` invocation. Not at the time we are resolving the extended type. For example, in ```AfterMembersChecks``` where we already perform some checks involving extended type.

Reply:
I agree with the first part. Moved this check to `AfterMembersChecks`.
For the second part (using a simpler `VisitType` invocation), I don't see how that would work, even if `VisitType` was enhanced to erase extension types.